### PR TITLE
Added CircularBuffer DataStructure.

### PR DESCRIPTION
**Pull request by NdubuisiJr**
Checkout the circular Buffer Data structure. Added unit tests too.

### CODE REVIEW

File: DataStructures/Lists/CircularBuffer.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+﻿using System;
+
+namespace DataStructures.Lists {
+    public class CircularBuffer<T> {
+        /// <summary>
+        /// Initializes a circular buffer with initial length of 10
+        /// </summary>
+        public CircularBuffer(bool canOverride=true) : this(10,canOverride) {
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: Can we add property and use it instead that hardcoded 10
        private static readonly int defaultBufferLength = 10;
This allow us avoid magic numbers in code.

File: DataStructures/Lists/CircularBuffer.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        /// <summary>
+        /// Initializes a circular buffer with given length
+        /// </summary>
+        /// <param name="length">The length of the buffer</param>
+        public CircularBuffer(int length, bool canOverride=true) {
+            if (length < 1) {
+                throw new ArgumentOutOfRangeException("length can not be zero or negative");
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: Maybe we should add test-case that cover that exception case?

File: DataStructures/Lists/CircularBuffer.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            else {
+                DontOverrides(value);
+            }
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer when it is not filled up
+        /// </summary>
+        /// <param name="value"></param>
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: I don't see point to add that empty param comment. I think your method naming is that good it doesn't required additional summary. I would leave only summary for public methods, private doesn't require this.

File: DataStructures/Lists/CircularBuffer.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            if (IsFilledUp) {
+                return;
+            }
+            InsertData(value);
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer without checking if it is full
+        /// </summary>
+        /// <param name="value"></param>
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: Same as above.

File: DataStructures/Lists/CircularBuffer.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads value from the front of the buffer
+        ///     <summay>
+        ///         Use the IsEmpty property to know when to read appropriately.
+        ///     </summay>
+        /// </summary>
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: What effect the extra inside summary block has? Can we safe remove that?

File: DataStructures/Lists/CircularBuffer.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        /// <returns></returns>
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: Return same as above comments, just remove or fill.

File: DataStructures/Lists/CircularBuffer.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        public int Length { get => _circularBuffer.Length - 1; }
+        public bool IsFilledUp { get => ((_end + 1) % _circularBuffer.Length == _start) && !_circularBuffer[_start].Equals(_circularBuffer[_end]); }
+        /// <summary>
+        /// Controls whether data should be overridden when it is continously inserted without reading
+        /// </summary>
+        public bool CanOverride { get; }
+
+        private T[] _circularBuffer;
+        private int _end;
+        private int _start;
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: I think more intuitive is put properties at begin of class, but it's not required to change, up to you.

File: DataStructures/Lists/CircularBuffer.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer when it is not filled up
+        /// </summary>
+        /// <param name="value"></param>
+        private void DontOverrides(T value) {
+            if (IsFilledUp) {
+                return;
```
Comment (PatrykOlejniczak) [inline DataStructures/Lists/CircularBuffer.cs]: Maybe we should throw exception (ArgumentOutOfRange  in this case? But here i am not sure, so open for discussion.

From use-case perspective if i add some property i expect it'll be added or i get some information back (if it won't be added), in this case it can be exception.

File: UnitTest/DataStructuresTests/CircularBufferTest.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace UnitTest.DataStructuresTests {
+    public class CircularBufferTest {
+        
+        [Fact]
+        public static void SetsFixedLength() {
+            var circularBuffer = new CircularBuffer<byte>(3);
+            var length = circularBuffer.Length;
+
+            Assert.Equal(3, length);
```
Comment (PatrykOlejniczak) [inline UnitTest/DataStructuresTests/CircularBufferTest.cs]: Can you add also asserts for properties IsEmpty and IsFilledUp  to be sure it's correctly set?

File: DataStructures/Lists/CircularBuffer.cs (hunk 1 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Yes we can. I added a private static readonly field to hold the default buffer length. Since properties are meant to be public.

File: DataStructures/Lists/CircularBuffer.cs (hunk 2 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Yeah, I think so too.

Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Added a test case that checked for length of 0 and -2. The test case was GREEN.

File: DataStructures/Lists/CircularBuffer.cs (hunk 3 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: I wanted to leave a helper comment for anyone who wants to go through the implementation. I guess a single line comment is more suitable. I have wiped out the summary.

File: DataStructures/Lists/CircularBuffer.cs (hunk 5 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Just figured that it doesn't have much to offer. I have removed it.

File: DataStructures/Lists/CircularBuffer.cs (hunk 6 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Removed

File: DataStructures/Lists/CircularBuffer.cs (hunk 7 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Since it's not part of our convention here, I will prefer to leave it that way. I have a personal arrangement for writing members of a class. Constructors, methods, events, properties and fields.

File: DataStructures/Lists/CircularBuffer.cs (hunk 8 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Throwing an exception will make more sense than just leaving the user with no clue as to what happened to the data he tried to insert. The custom exception name is excellent. I've implemented it.

File: UnitTest/DataStructuresTests/CircularBufferTest.cs (hunk 9 above)
Comment (NdubuisiJr) [reply UnitTest/DataStructuresTests/CircularBufferTest.cs]: Sure

Comment (NdubuisiJr) [reply UnitTest/DataStructuresTests/CircularBufferTest.cs]: I have added two test cases; CheckIsEmptyProperty and CheckIsFilledProperty, both test cases are GREEN.

File: DataStructures/Lists/CircularBuffer.cs (hunk 8 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: I also added test cases to confirm that the exception is throw appropriately.

**Comment by NdubuisiJr** [comment]
changed the '{' to match the repo's convention. Implemented our suggestions on the review comments.

File: UnitTest/DataStructuresTests/CircularBufferTest.cs [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        [Fact]
+        public static void TestingCantOverrideFunctionality() 
+        {
+            var circularBuffer = new CircularBuffer<byte>(3, false);
+            circularBuffer.Write(3);
+            circularBuffer.Write(34);
+            circularBuffer.Write(24);
+            // if it doesn't override, then the values above will be read else, 
+            // the value below will be read
+            circularBuffer.Write(2);
```
Comment (PatrykOlejniczak) [inline UnitTest/DataStructuresTests/CircularBufferTest.cs]: This test will fail because CanOverride is false and we try input more than buffer limit.

Comment (NdubuisiJr) [reply UnitTest/DataStructuresTests/CircularBufferTest.cs]: Yeah... because of the custom exception we are throwing in the DontOverrides() method when not overriding.

Comment (NdubuisiJr) [reply UnitTest/DataStructuresTests/CircularBufferTest.cs]: So, I added an assertion to show that the exception is thrown. I also left the assertions that ensures the values in the buffer are correct.

**Comment by NdubuisiJr** [comment]
Fixed failing unit test

**Comment by PatrykOlejniczak** [comment]
Thanks, looks good, great job!

File: DataStructures/Lists/CircularBuffer.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            if (IsFilledUp) 
+            {
+                throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+            }
+            InsertData(value);
+        }
+
+        // Inserts data into the buffer without checking if it is full
+        private void InsertData(T value)
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Please use `thisCase` for private methods and `ThatCase` for public methods. It will help people visually understand the difference. Also, it would make more sense if you renamed the method to `innerInsert` than `InsertData`, as the `Data` word here is not needed.

File: DataStructures/Lists/CircularBuffer.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        /// </summary>
+        public bool CanOverride { get; }
+        public bool IsEmpty { get => _end == _start; }
+        public int Length { get => _circularBuffer.Length - 1; }
+        public bool IsFilledUp { get => ((_end + 1) % _circularBuffer.Length == _start) && !_circularBuffer[_start].Equals(_circularBuffer[_end]); }
+
+        private T[] _circularBuffer;
+        private int _end;
+        private int _start;
+        private static readonly int _defaultBufferLength = 10;
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Please move all instance variables, private and public, to the beginning of the class, right before the constructor. It should be the first thing that a contributor should see when viewing your code. Also, check the other data structures to see how variables are ordered.

File: DataStructures/Lists/CircularBuffer.cs [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            if (_end == _start) 
+            {
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads value from the front of the buffer
+        /// </summary>
+        public T Read()
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Maybe a better name for this method is `Pop`, since it mutates the buffer? `Read` makes it sound as if the method is side-effect-free, where in fact it mutates the internal state of the buffer. Also please extend the documentation to reflect this, saying that the method just reads is misleading.

File: DataStructures/Lists/CircularBuffer.cs [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+            _end = 0;
+            _start = 0;
+            CanOverride = canOverride;
+        }
+
+        /// <summary>
+        /// Writes value to the back of the buffer
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Write(T value)
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: The use of the `Write` word is not appropriate, if I use the `List.Write()` method, it will make me believe I am re-writing the entire lists contents with that argument, a more suitable name would be `Add`, since that's what the method is doing.

**Comment by aalhour** [comment]
Please refer to this Wikipedia article, specially the reference to Read and Write pointers: [Circular Buffer](https://en.wikipedia.org/wiki/Circular_buffer).

File: DataStructures/Lists/CircularBuffer.cs (hunk 11 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Okay. 
Done

File: DataStructures/Lists/CircularBuffer.cs (hunk 12 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: The way it is right now is a result of my personal convention. Guess I'll have to stick to the Repo's convention :) 
Done.

File: DataStructures/Lists/CircularBuffer.cs (hunk 13 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Done

File: DataStructures/Lists/CircularBuffer.cs (hunk 14 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Yeah. It makes absolute sense.

Done.

File: DataStructures/Lists/CircularBuffer.cs [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+        // Inserts data into the buffer when it is not filled up
+        private void DontOverrides(T value) 
+        {
+            if (IsFilledUp) 
+            {
+                throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+            }
+            InsertData(value);
+        }
```
[the remark this answers is not in the document]
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: That's what I've done

**Comment by NdubuisiJr** [comment]
Concerning the Read and Write pointers. The additional mechanism used for checking the positions of the start and end indexes when using two pointers talked about in this article [https://en.wikipedia.org/wiki/Circular_buffer](url). The article suggested that the buffer should to hold (size - 1) items. That is what I actually did In the constructor when `_circularBuffer = new T[length + 1];`. But Instead of decreasing the overall size I just increased it by 1. The logic works the same way.

**Comment by aalhour** [comment]
> Concerning the Read and Write pointers. The additional mechanism used for checking the positions of the start and end indexes when using two pointers talked about in this article [Circular_buffer](https://en.wikipedia.org/wiki/Circular_buffer). The article suggested that the buffer should hold (size - 1) items. That is what I actually did In the constructor when instantiating the array `_circularBuffer = new T[length + 1];`. But Instead of decreasing the overall size I just increased it by 1. The logic works the same way.

Thanks for implementing the changes, the code looks much better now 👍 

The reason I referenced the article, which maybe I should have made more explicit, is the fact that a Circular Buffer would allow you to manipulate the write and read indexes separately which will make it a building block for other circular data structures, i.e.: Circular Queues and Circular Stacks, but since you don't provide that in your implementation we will have to make it a higher level abstraction, i.e.: Circular Stack, or we agree that you need to implement a lower level Circular Buffer which would then be used in other Circular Linear Collection implementations (CQueues and CStacks). Btw, take a look at PR #80 which supposedly implements a Circular Queue, it won't be possible to have two implementations of the same exact thing.

File: DataStructures/Lists/CircularBuffer.cs [hunk 16]
```diff
@@ -0,0 +1,92 @@
+﻿using System;
+
+namespace DataStructures.Lists 
+{
+    public class CircularBuffer<T>
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Can you please implement the `IEnumerable` interface? Please refer to how the Linear Collections were implemented, e.g.: List, Stack, Queue ... etc.

Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: I've Implemented both.

**Comment by NdubuisiJr** [comment]
Done. Also, added a test method to very the methods in `ICollection `interface

File: DataStructures/Lists/CircularBuffer.cs [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            get { return false; }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            for (int i = 0; i < _circularBuffer.Length; i++) 
+            {
+                _circularBuffer[i] = _default;
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Why a for loop and not create a new one?
```
_circularBuffer = new T[length + 1];
```

File: DataStructures/Lists/CircularBuffer.cs [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+        /// Copies this buffer to an array
+        /// </summary>
+        public void CopyTo(T[] array, int arrayIndex) 
+        {
+            if (array == null) 
+            {
+                throw new ArgumentNullException("array can not be null");
+            }
+
+            if (array.Length == 0 || arrayIndex >= array.Length || arrayIndex < 0) {
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Inconsistent formatting, sometimes you place the `{` on the same line and sometimes you place it on a new line. Please place all `{`, `}` characters on new lines.

File: DataStructures/Lists/CircularBuffer.cs [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+                    break;
+                }
+            }
+        }
+        /// <summary>
+        /// Removes an item from the buffer
+        /// </summary>
+        public bool Remove(T item) 
+        {
+            if (Contains(item))
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: No need for the `Contains` method call, you are already looping through the collection, might as well handle the boolean check there. Also, `Contains` has an implementation of `O(n)`, which means with this implementation you're looping over the collection twice.

File: DataStructures/Lists/CircularBuffer.cs [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+                innerInsert(value);
+            }
+            else 
+            {
+                if (IsFilledUp) 
+                {
+                    throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+                }
+                innerInsert(value);
+            }
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: This method can be simplified with a single if-else clause, which would only contain a single call to `innerInsert` method, i.e.:
```
If(CanOverride == false && IsFilledUp == true)
{
    // throw new exception
}

innerInsert(value)
```

File: DataStructures/Lists/CircularBuffer.cs [hunk 21]
```diff
@@ -0,0 +1,197 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DataStructures.Lists 
+{
+    public class CircularBuffer<T> : IEnumerable<T> , ICollection<T> where T : IComparable<T> {
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Inconsistent formatting, sometimes you place the `{` on the same line and sometimes you place it on a new line. Please place all `{`, `}` characters on new lines.

File: DataStructures/Lists/CircularBuffer.cs [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+        }
+        #endregion
+
+        #region ICollection Implementation
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get { return _length; }
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Incorrect implementation. `Count` should return the number of inserted elements into the buffer not the length of the internal collection. The `_circularBuffer` could be empty and of length 10, but that doesn't mean that the count of elements it contains is 10. Please re-implement this based on an internal counter.

File: DataStructures/Lists/CircularBuffer.cs [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+        private int _start;
+        private int _length { get => _circularBuffer.Length - 1; }
+        private readonly T _default;
+        private static readonly int _defaultBufferLength = 10;
+
+        /// <summary>
+        /// Controls whether data should be overridden when it is continously inserted without reading
+        /// </summary>
+        public bool CanOverride { get; }
+        public bool IsEmpty { get => _end == _start; }
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: Does this make sense if it's a circular buffer with a `CanOverride=true` case? What would `IsEmpty` mean here?

File: DataStructures/Lists/CircularBuffer.cs (hunk 17 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Implemented

File: DataStructures/Lists/CircularBuffer.cs (hunk 18 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Sorry for that. I came from the Java side of things. It's all consistent now

Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Sorry about that. I came from the Java side of things. It's all consistent now

File: DataStructures/Lists/CircularBuffer.cs (hunk 20 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: This refactoring is ingeneous. Wonder why I didn't think of it :)

File: DataStructures/Lists/CircularBuffer.cs (hunk 19 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Please review how it is now.

File: DataStructures/Lists/CircularBuffer.cs (hunk 23 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Please review the new Implementation. added a `_count` field that monitored each insertion, pop, and Removal of elements.

File: DataStructures/Lists/CircularBuffer.cs (hunk 22 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Please review the new implementation.

**Comment by NdubuisiJr** [comment]
Please Review the current state

File: DataStructures/Lists/CircularBuffer.cs (hunk 22 above)
Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: Looks better.

File: DataStructures/Lists/CircularBuffer.cs (hunk 19 above)
Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: Looks better.

**Comment by aalhour** [comment]
@NdubuisiJr - the data structure looks much better and more polished now. I have one remaining question though. The `Remove` method allows the user to remove a single element from the buffer, which would leave the buffer with a non-existent extra item assigned to the same slot in the array (`default(T)`), I think this might be a problem, here is why:

Let's say the buffer contains 5 elements
```
|  1  |  2  |  3  |  4  |  5  |
```

And then, I choose to delete `4`, the buffer will assign the default value of Integers, which is `0`, in that slot:
```
|  1  |  2  |  3  |  *0*  |  5  |
```

I think a better implementation of `Remove` would actually shift all element after the one you just removed one slot to the back, and then update the `_end` pointer to reference the new slot where the end element is, leaving buffer with an open slot at the end:

```
```
|  1  |  2  |  3  |  5  |  *0*  |
```

I think this will leave us with a better implementation but will make things hairier, i.e.:
 * Should we move all elements, after the deleted item, back one spot?
  * Or should we just construct a new array and use the `Array.copyTo` API to copy things over and then discard the existing one?

What do you think?

**Comment by NdubuisiJr** [comment]
@aalhour, I really want to appreciate you and @PatrykOlejniczak for the reviews so far. I've really learned much from you guys.

I think we should go for the first option you gave because it reduces the number of loops and `if` checks.

**Comment by NdubuisiJr** [comment]
Used a while loop to move each element after the deleted one.

Please review the current state. 
Thanks.

File: DataStructures/Lists/CircularBuffer.cs [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+                        var temp = _circularBuffer[j];
+                        _circularBuffer[j] = _circularBuffer[j + 1];
+                        _circularBuffer[j + 1] = temp;
+                        j++;
+                    }
+                    _end = i + _count ;
+                    break;
+                }
+            }
+            return result;
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: I think the following is a much better solution for removing an element from an array:
```csharp
# Assume _circularBuffer is the following: { 1, 3, 4, 9, 2 }
if (Contains(item))
{
    _circularBuffer = _circularBuffer.Except(new T[]{ item }).ToArray();
    
    // Make sure that `_count` and `_end` remain within valid bounds 
    if (Count > 0)
    {
        --_count;
        --_end;
    }
    
    return true;
}

return false;
```

Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: Please add test cases for deleting the `default(T)` value from an empty buffer.

Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: Actually, the code I wrote above has a bug of incorrectly calculating the value for `_end` pointer. Your code also has that bug. Here's how to reproduce:

Let's say you have an almost full buffer of length 5 (last spot remains):
```
var buffer = new int[]{ 11, 12, 13, 14 }
```

At this point the `_start` pointer is equal to `0` and the `_end` pointer is equal to `4`, because: `(3 + 1) % 5 = 4`, which is the expression that gets evaluated every time you insert a new element.

What would happen if we insert a new element of value `15`? The buffer will look like:
```
{ 11, 12, 13, 14, 15}
```

And the `_end` pointer will be set to `0`, because: `(4 + 1) % 5 = 0`.

So, what would happen if we were to delete the item `14`? The buffer will look like the following:

```
{ 11, 12, 13, 15, <EMPTY> }
```

But the `_end` pointer will be set to `-1`, because the `Remove` method just decrements it. I think the `_end` pointer needs to be recalculated based on the `_count` value, right before the `_count` gets decreased, as follows:
```csharp
// Recalculate _end pointer value
_end = (Count - 1) % Length;

// Decrease the _count value now
--_count;
```

In the case of  a full buffer where `_end = _start = 0`, the `_end` pointer will be re-calculated to `4`, because: `(5 - 1) % 5 = 4`. This method scales for every size of the buffer, here are the calculations:
  * Full buffer: `_end = (5 - 1) % 5;    // 4 (5th index of the array, currently free)`
  * 4/5 full buffer: `_end = (4 - 1) % 5;    // 3 (Second-last index of the array, currently free)`
  * 3/5 full buffer: `_end = (3 - 1) % 5;    // 2 (3rd / mid index of the array, currently free)`
  * 2/5 full buffer: `_end = (2 - 1) % 5;    // 1 (2nd index of the array, currently free)`
  * 1/5 full buffer: `_end = (1 - 1) % 5;    // 0 (1st index of the array, empty buffer)`

Please review what I just wrote and let me know if anything is not clear.

Also, please document the order of operations in the new implementation, why we re-calculate the `_end` pointer first before decrementing the `_count` ... etc.

Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: Please add tests to cover the following:
  * Try removing the `default(T)` value from the buffer and make sure the buffer is not affected as long as the elements don't contain that value
  * Try adding duplicate entries to the buffer and making sure all of them get deleted when the `Remove(T item)` gets called and that the difference in buffer length is equal to the original length minus the number of duplicate entries that were in the buffer, originally
  * Removing elements one by one from the end of the buffer to the front, results in a correct state of the buffer and when inserting them back in the buffer looks identical to how it was before the items were deleted

Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Once, I added these tests to my implementation, the bug jumped out :) . Yours is quite simple. Let me implement it and see how it goes.

Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: After implementing your suggested code. The first test case passed but the other two failed. I found some other bugs:

-This code  `_circularBuffer = _circularBuffer.Except(new T[]{ item }).ToArray();` shrinkes the `Length` of the buffer. The buffer is meant to be a fixed "Ring" (have fixed length) but the element count can change.

- When we remove identical elements, say we had a buffer that looks like this `{1,2,3,3,3}`, The count is only decreased by one because of this line `--_count;`, whereas it's meant to decrease by three because only two elements will be left: `{1,2}`.

- The point above also affects the `_end` pointer. Instead of decreasing by 3, it's being decreased by one because of this line `_end = (Count - 1) % Length;`.

Following your line of thought, this is what I think we should implement:
`if (!IsEmpty && Contains(item)) 
            {
               var sourceArray = _circularBuffer.Except(new T[] { item }).ToArray();
                _circularBuffer = new T[Length + 1];
                Array.Copy(sourceArray, _circularBuffer, sourceArray.Length);

                if (!item.Equals(default(T))) 
                {
                    _end = sourceArray.Length - 1;
                    _count = sourceArray.Length-1;
                }  
                return true;
            }

            return false;`
I have Implemented this on my local machine and all the test cases passed. To explain how it solved the problem, let's create a buffer with 4 elements. To the client, it looks like this: `{1,2,3,3}` but to the `CircularBuffer<T> class` it looks like this `{1,2,3,3,<empty>}`. This <empty> space is required when using two pointers as discussed in this Wikipedia article [Circular_buffer](https://en.wikipedia.org/wiki/Circular_buffer). So,

- When we remove `default(T)`, which is equivalent to `<empty>`, the `sourceArray` will look like this `{1,2,3,3}`. we now create a new `_circularBuffer`, with our initial Length. We then copy the   `sourceArray` into our empty  `_circularBuffer`. which now looks like the original - `{1,2,3,3,<empty>}`.  Since we are removing `default(T)`, or _end and _count should not change - that's what the `if` statement helps us do.

- If we remove 3 the `sourceArray` will look like this `{1,2,<empty>}`. Thus, by copying it to the newly created `_circularBuffer` we now have `{1,2,<empty>,<empty>,<empty>}`. we now have to recalculate the `_end` pointer as `_end = sourceArray.Length - 1;` (insert at index 2). The `_count` is also recalculated as `_count=sourceArray.Length-1` (equals 2).

Please review the algorithm. Once we resolve her, I'll push the changes :) . Thanks

Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: @NdubuisiJr, good catch! Your line of thought looks tight, please push your changes to this PR. I will mark this conversation as resolved! I am looking forward to see an extensive test coverage for this method.

Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Okay here goes...

File: DataStructures/Lists/CircularBuffer.cs [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+        /// </summary>
+        public bool Remove(T item) 
+        {
+            if (!IsEmpty && Contains(item)) 
+            {
+               var sourceArray = _circularBuffer.Except(new T[] { item }).ToArray();
+                _circularBuffer = new T[Length + 1];
+                Array.Copy(sourceArray, _circularBuffer, sourceArray.Length);
+
+                if (!item.Equals(default(T)))
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: I think this `if` statement is not needed, but I might be wrong, so can you please add a test for removing duplicate `default(T)` values from a buffer? Here's an example case:

```
{ 1, 0, 2, 0, 3, 0, 4, 0, 5, *0*, *0*, *0* }
```

Where, `*0*` is an empty slot, but `0` is not, and was actually added by the user.

File: DataStructures/Lists/CircularBuffer.cs [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+            if (!IsEmpty && Contains(item)) 
+            {
+               var sourceArray = _circularBuffer.Except(new T[] { item }).ToArray();
+                _circularBuffer = new T[Length + 1];
+                Array.Copy(sourceArray, _circularBuffer, sourceArray.Length);
+
+                if (!item.Equals(default(T))) 
+                {
+                    _end = sourceArray.Length - 1;
+                    _count = sourceArray.Length-1;
```
Comment (aalhour) [inline DataStructures/Lists/CircularBuffer.cs]: `_count` calculation is wrong. Count *IS* the length of the `sourceArray`, since that contains the remaining elements.

File: DataStructures/Lists/CircularBuffer.cs (hunk 25 above)
Comment (aalhour) [reply DataStructures/Lists/CircularBuffer.cs]: Please refer to the following REPL, I think there are several problems with the new implementation of `Remove()`, please extend the test cases to also cover removing duplicate entries that are `default(T)` for both integers and strings.

REPL: https://repl.it/@aalhour/csharp-circular-buffer-pr-review

File: DataStructures/Lists/CircularBuffer.cs (hunk 26 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Apologies for not replying earlier.
The `_count` calculation is not wrong. Remember that for the case where `item` is not equal to `default(T)`, the `sourceArray` will be reduced to `{[values], <empty>}`. So, the `_count` is the length of the [values] minus the <empty> space.

File: DataStructures/Lists/CircularBuffer.cs (hunk 25 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Apologies for not replying earlier.

You are absolutely right, we need to update the `_end `and `_count `all the time. But, we need the `if` statement. Actually we need an `if-else` statement ( at least, I think it's easier to understand ).

File: DataStructures/Lists/CircularBuffer.cs (hunk 26 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Yes, it is wrong when we are removing `default(T)`, because the length of the `sourceArray` exactly matches the number of elements left. But, it is correct when we are removing other values, because the `sourceArray` will contain `{[other values],<empty>}`. Therefore the count should be one less than the length of `sourceArray`.

File: DataStructures/Lists/CircularBuffer.cs (hunk 25 above)
Comment (NdubuisiJr) [reply DataStructures/Lists/CircularBuffer.cs]: Here's how I modified it currently.
                ```
if (!Equals(item,default(T))) 
                {
                    _end = sourceArray.Length - 1;
                    _count = sourceArray.Length-1;
                }
                else 
                {
                    _end = sourceArray.Length;
                    _count = sourceArray.Length;
                }
```
Lets consider a buffer of length 12:  `{1,0,2,0,3,0,4,0,5,*0*,*0*,*0*,*0*}`. Internally, it's of length 13 as explained earlier.

When we remove `0`, `sourceArray` becomes `{1,2,3,4,5}` and `_circularBuffer` becomes {1,2,3,4,5,*0*,*0*,*0*,*0*,*0*,*0*,*0*,*0*}. This clearly shows that `sourceArray.Length` is the correct `_count` of `_circularBuffer`. The next value should be inserted at `index=5`. Thus `_end` is still equal to the `sourceArray.Length`.

However, when we are removing item that is not equal to `default(T)`, these calculations have to change because of the `<empty space>` that will be left in the `sourceArray`.

That's why I used the `if-else` construct. This implementation currently passes all the tests including that of strings. Please review.
